File: src/main/java/org/zenbot/szolnok/timetable/batch/step/bus/processor/JsoupDocumentToTimetableProcessor.java
package org.zenbot.szolnok.timetable.batch.step.bus.processor;

import lombok.extern.slf4j.Slf4j;
import org.jsoup.nodes.Document;
import org.springframework.batch.item.ItemProcessor;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.stereotype.Component;
import org.zenbot.szolnok.timetable.configuration.properties.TimetableProperties;
import org.zenbot.szolnok.timetable.configuration.properties.TimetableSelectorProperties;
import org.zenbot.szolnok.timetable.domain.Timetable;

import java.util.Map;

@Slf4j
@Component
@EnableConfigurationProperties(TimetableProperties.class)
public class JsoupDocumentToTimetableProcessor implements ItemProcessor<Document, Timetable> {

    public static final String WEEKDAY_KEY = "weekday";
    public static final String SATURDAY_KEY = "saturday";
    public static final String SUNDAY_KEY = "sunday";

    private final TimetableSelectorProperties selectorProperties;

    private final EndBusStopSelectorItemProcessorHelper endBusStopSelectorItemProcessorHelper;
    private final ActualStopSelectorItemProcessorHelper actualStopSelectorItemProcessorHelper;
    private final TimetableRowBuilderItemProcessorHelper timetableRowBuilderItemProcessorHelper;

    public JsoupDocumentToTimetableProcessor(TimetableProperties properties, EndBusStopSelectorItemProcessorHelper endBusStopSelectorItemProcessorHelper, ActualStopSelectorItemProcessorHelper actualStopSelectorItemProcessorHelper, TimetableRowBuilderItemProcessorHelper timetableRowBuilderItemProcessorHelper) {
        this.selectorProperties = properties.getSelector();
        this.endBusStopSelectorItemProcessorHelper = endBusStopSelectorItemProcessorHelper;
        this.actualStopSelectorItemProcessorHelper = actualStopSelectorItemProcessorHelper;
        this.timetableRowBuilderItemProcessorHelper = timetableRowBuilderItemProcessorHelper;
    }

    @Override
    public Timetable process(Document htmlDocument) {
        Timetable timetable = new Timetable();
        String busName = htmlDocument.select(selectorProperties.getRouteNameSelector()).text();
        String startBusStop = htmlDocument.select(selectorProperties.getFromSelector()).text();
        String endBusStop = endBusStopSelectorItemProcessorHelper.getEndBusStop(htmlDocument, selectorProperties);
        String actualStop = actualStopSelectorItemProcessorHelper.getActualStop(htmlDocument, selectorProperties);
        Map<Integer, Map<String, String>> timetableRows = timetableRowBuilderItemProcessorHelper.getTimetableRows(htmlDocument, selectorProperties);

        timetable.setBusName(busName);
        timetable.setStartBusStopName(startBusStop);
        timetable.setEndBusStopName(endBusStop);
        timetable.setActiveStopName(actualStop);
        timetableRows.forEach(timetable::addRow);

        log.info("Process html with busName [#{}] and timetable for [{}] stop", timetable.getBusName(), timetable.getActiveStopName());
        return timetable;
    }

}

File: src/main/kotlin/org/zenbot/szolnok/timetable/batch/step/bus/processor/JsoupDocumentToTimetableProcessor.kt
package org.zenbot.szolnok.timetable.batch.step.bus.processor

import org.jsoup.nodes.Document
import org.slf4j.LoggerFactory
import org.springframework.batch.item.ItemProcessor
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.stereotype.Component
import org.zenbot.szolnok.timetable.configuration.properties.TimetableProperties
import org.zenbot.szolnok.timetable.configuration.properties.TimetableSelectorProperties
import org.zenbot.szolnok.timetable.domain.Timetable

@Component
@EnableConfigurationProperties(TimetableProperties::class)
class JsoupDocumentToTimetableProcessor(properties: TimetableProperties, private val startBusStopSelectorItemProcessorHelper: StartBusStopSelectorItemProcessorHelper, private val endBusStopSelectorItemProcessorHelper: EndBusStopSelectorItemProcessorHelper, private val actualStopSelectorItemProcessorHelper: ActualStopSelectorItemProcessorHelper, private val timetableRowBuilderItemProcessorHelper: TimetableRowBuilderItemProcessorHelper) : ItemProcessor<Document, Timetable> {

    private val log = LoggerFactory.getLogger(JsoupDocumentToTimetableProcessor::class.java)

    private val selectorProperties: TimetableSelectorProperties

    init {
        this.selectorProperties = properties.selector
    }

    override fun process(htmlDocument: Document): Timetable {
        val timetable = Timetable()
        val busName = htmlDocument.select(selectorProperties.routeNameSelector).text()
        val startBusStop = startBusStopSelectorItemProcessorHelper.getStartBusStop(htmlDocument, selectorProperties)
        val endBusStop = endBusStopSelectorItemProcessorHelper.getEndBusStop(htmlDocument, selectorProperties)
        val actualStop = actualStopSelectorItemProcessorHelper.getActualStop(htmlDocument, selectorProperties)
        val timetableRows = timetableRowBuilderItemProcessorHelper.getTimetableRows(htmlDocument, selectorProperties)

        timetable.busName = busName
        timetable.startBusStopName = startBusStop
        timetable.endBusStopName = endBusStop
        timetable.activeStopName = actualStop
        timetableRows.forEach { hour, values -> timetable.addRow(hour, values) }

        log.info("Process html with busName [#{}] and timetable for [{}] stop", timetable.busName, timetable.activeStopName)
        return timetable
    }

    companion object {

        val WEEKDAY_KEY = "weekday"
        val SATURDAY_KEY = "saturday"
        val SUNDAY_KEY = "sunday"
    }
}
